Ext/ThesisPeter/LightNodeManager.h

```diff
@@ -0,0 +1,380 @@
+			throw std::invalid_argument(unimplementedFormatMsg + name.toString() + '\'');
+		}
+	}
+
+	virtual ~LightNodeIndexAttributeAccessor(){}
+
+	const unsigned int getLightNodeIndex(uint32_t index) const {
+		assertRange(index);
+		const float* v = _ptr<const float>(index);
+		return *v;
+	}
+
+	void setLightNodeIndex(uint32_t index, const float value){
+		assertRange(index);
+		float* v = _ptr<float>(index);
+		*v = value;
+	}
+};
+}
+
+namespace MinSG {
+class Node;
+class GeometryNode;
+class LightNode;
+class PathNode;
+
+namespace ThesisPeter {
+
+struct FilterEdgeState {
+	int childID;
+	int nodeOffset;
+	Geometry::Vec3 t0, t1, tm;
+};
+
+struct LightNode {
+	Geometry::Vec3 osPosition;	//object space position, used for dynamic objects for recalculation of the position
+	Geometry::Vec3 position;
+	Geometry::Vec3 osNormal;	//object space normal, used for dynamic objects for recalculation of the positions
+	Geometry::Vec3 normal;
+	Util::Color4f color;
+	unsigned int id;			//used to identify the position inside the texture on GPU
+};
+
+struct LightEdge {
+	LightNode* source;
+	LightNode* target;
+	Geometry::Vec3 weight;
+};
+
+struct LightNodeMap;
+
+struct LightNodeMapConnection {
+	std::vector<LightEdge*> edges;
+	LightNodeMap* map1;
+	LightNodeMap* map2;
+};
+
+struct LightNodeMap {
+	std::vector<LightNode*> lightNodes;
+	std::vector<LightEdge*> internalLightEdges;
+	std::vector<LightNodeMapConnection*> externalLightEdgesStatic;
+	std::vector<LightNodeMapConnection*> externalLightEdgesDynamic;
+	MinSG::GeometryNode* geometryNode;
+	bool staticNode;
+};
+
+struct LightNodeLightMap {
+	LightNode light;
+	MinSG::LightNode* lightNode;
+	std::vector<LightEdge*> staticEdges;
+	std::vector<LightEdge*> dynamicEdges;
+	bool staticNode;
+};
+
+struct LightingArea {
+	float extend;
+	Geometry::Vec3 center;
+};
+
+class LightInfoAttribute : public Util::GenericAttribute {
+public:
+	unsigned int lightNodeID;
+	bool staticNode;
+
+	LightInfoAttribute* clone() const override {
+		LightInfoAttribute* lightAttr = new LightInfoAttribute();
+		lightAttr->lightNodeID = lightNodeID;
+		lightAttr->staticNode = staticNode;
+		return lightAttr;
+	}
+};
+
+class NodeCreaterVisitor : public MinSG::NodeVisitor {
+public:
+	NodeCreaterVisitor();
+	static unsigned int nodeIndex;
+	std::vector<LightNodeMap*>* lightNodeMaps;
+	std::vector<LightNodeLightMap*>* lightNodeLightMaps;
+	std::vector<MinSG::PathNode*>* pathNodes;
+	bool useGeometryNodes, useLightNodes, loadPaths;
+
+private:
+	static const Util::StringIdentifier staticNodeIdent;
+
+	status leave(MinSG::Node* node);
+};
+
+class NodeRenderVisitor : public MinSG::NodeVisitor {
+public:
+	NodeRenderVisitor();
+
+	void init(MinSG::FrameContext* frameContext);
+
+private:
+	status leave(MinSG::Node* node);
+
+	MinSG::FrameContext* frameContext;
+	MinSG::RenderParam renderParam;
+};
+
+class LightNodeManager {
+	PROVIDES_TYPE_NAME(LightNodeManager)
+public:
+	LightNodeManager();
+	virtual ~LightNodeManager();
```

Comment:
does it matter though? At least if someone would derive from this class it would behave as expected in most cases, I guess ...